Vignette update

pat-s opened the pull request:
#78

krlmlr commented:
Deploy preview for *wonderful-volhard-261e08* ready!

Built with commit 6ba8e8189f35bdfae5f0858e9cae4b4acc497965

https://deploy-preview-80--wonderful-volhard-261e08.netlify.com

krlmlr commented:
Thanks! I'll tweak and update this branch tomorrow, and also provide feedback to the questions in #78.

krlmlr commented:
Thanks. I've read the text again, I like it. I tweaked some typos, could you please review my changes in  d00d1ef3865bd9e899e343ea9d5456612650d03b and 08c5172aea3815b47a0f618755f4bc6d62afb9c1, and address the following minor remarks:

- [ ] its -> it is | it's
- [ ] Please run spell checker!
- [ ] Formatting for packages: _tic_  vs. `tic`

# Get started

- [ ] Familiarity with version control systems taken for granted?
- [ ] Use new structure from README: repeat three advantages, include minor advantages (Robust caching, package checking with rcmdcheck)

# Advantages

- [ ] Use new structure from README: repeat three advantages, include minor advantages (Robust caching, package checking with rcmdcheck)
    - Add subsection "Support for various projects"
    - Reorder subsections
    - Make sure overview is consistent with subsections

- [ ] Caching: No more multiple libraries (sorry!)


# Build lifecycle

- [ ] Diagrams: Great to have the sources in the repo. How to convert `.graphml` files? Please add README or `Makefile`.

- [ ] Please rephrase, perhaps split sentence:
    > `tic` also uses the stages approach and simply fowards all the commands that should be run in stage to the `tic.R` file where these will specified in an CI-agnostic way using R syntax.

# Advanced

- [ ] When conditioning on R version, use `getRversion()` also in text. Double-check < and >.

- [ ] Pretty-print contents of stage

- [ ] "Troubleshooting: Running tic locally"
    - Introduce two subsections
    - Drop reference to earlier article
    - Add mini-code-block with `tic::tic()` call

# Deployment

- [ ] "Avoiding duplicate site creations in build matrices": If there's no practical difference, why spend so much space?
    - Practical difference: if _tic_ thinks it needs to deploy, it will install the build-time dependencies even if `tic::deploy()` is never called.
    - Rather focus on explaining the `tic.R` approach

pat-s commented:
> Familiarity with version control systems taken for granted?

I would say yes, VC comes before CI. I expect 99% of users interested in CI to be familiar with VC

pat-s commented:
Continuing in the next two days, not yet finished

pat-s commented:
> Pretty-print contents of stage

Even when setting 

    ```{r, comment="", results="asis", echo = FALSE}
    options(crayon.enabled = TRUE)
    old.hooks <- fansi::set_knit_hooks(knitr::knit_hooks)
    ```

the code block the output is not colored. Any idea why?

pat-s commented:
- Theme: cosmo
- Inline ToC
- Added example of non-standard tic based solutions
- added example output of rcmdcheck() failure
- syntax highlighting of code
- consistent italics for package names

krlmlr commented on vignettes/advanced.Rmd:
```diff
@@ -221,20 +112,174 @@ See `?ci` for more information on which information can be extracted from this f
+If your Travis build fails and you have no clue why (error messages are too unspecific, you cannot reproduce the problem locally), you can enter into the Travis build via SSH.
+To do so, your repository must be enabled for "ssh debugging".
+This means you need to write the Travis CI team an e-mail, asking whether your repo can be [activated for debugging](https://docs.travis-ci.com/user/running-build-in-debug-mode/#enabling-debug-mode).
+Usually you get a reply within one day with some additional information on how to enter the build.
```
"enter" is overused in this paragraph.

Also, I'm missing some detail on how this works:

1. Need to pick a job we want to troubleshoot.
2. Need to restart it (with our helper).
3. Need to SSH into it.

krlmlr commented on vignettes/advanced.Rmd:
````diff
@@ -221,20 +112,174 @@ See `?ci` for more information on which information can be extracted from this f
+travis_run_after_script
+```
+
+The debug session is a VM running a [`tmate`](https://tmate.io/) session.
````
Can we mention that `tmux` is running, and point to https://tmuxcheatsheet.com/?

pat-s replied:
done in https://github.com/ropenscilabs/tic/pull/80/commits/c6507532ae6fc36f475f39264575253a01e01337

krlmlr commented on vignettes/advanced.Rmd:
```diff
@@ -221,20 +112,174 @@ See `?ci` for more information on which information can be extracted from this f
 
-Once you are in the build [check](https://docs.travis-ci.com/user/running-build-in-debug-mode/#Things-to-do-once-you-are-inside-the-debug-VM) these handy functions to execute specific stages.
+- `SystemRequirements`: *tic* is not yet capable of automatically determining system requirements specified in DESCRIPTION files of an R package. 
+Our plan is to automatically provide suggestions like "Please add addons: apt: <library> to `.travis.yml`".
```
```suggestion
Our plan is to automatically provide suggestions like "Please add `addons: apt: <library>` to `.travis.yml`".
```

krlmlr commented on vignettes/advantages.Rmd:
```diff
@@ -0,0 +1,176 @@
+This means that these services can push certain files created during the build to repositories (Github, Gitlab, etc.).
+A common use case is to build your `pkgdown` site and conditionally push changes to your repo.
+This practice saves you from building the whole `pkgdown` site locally and then committing manually all the changed files - either Travis or Appveyor will do it for you! 
+*tic* takes care of all the setup work (setting up SSH keys, permission granting for the repo on Github and the CI services).
```
Maybe bullet points give us more space here? We could also mention setup of PATs.

(permission granting -> granting permission)

pat-s replied:
https://github.com/ropenscilabs/tic/pull/80/commits/d0806a8119e97023ab50c0d5027f064efb3d4836

krlmlr commented on vignettes/advantages.Rmd:
```diff
@@ -0,0 +1,176 @@
+
+## Support for various R projects
+
+*tic* was developed with the aim to not only simplify R package development using CI services but also to support other R project types (bookdown, blogdown, drat, website and fighshare deployment). 
```
```suggestion
*tic* was developed with the aim to not only simplify R package development using CI services but also to support other R project types (bookdown, blogdown, drat, website and figshare deployment). 
```

krlmlr commented on vignettes/advantages.Rmd:
```diff
@@ -0,0 +1,176 @@
+What does "CI-Agnostic" mean and why do we need it?
+
+You may have heard that multiple CI services exist in the wild.
+Some only support specific operating systems or certain languages and for some you need to pay a monthly rate to use them.
```
Not sure:

```suggestion
Some only support specific operating systems or certain languages and for some you need to pay a monthly fee to use them.
```

krlmlr commented on vignettes/advantages.Rmd:
```diff
@@ -0,0 +1,176 @@
+
+You may have heard that multiple CI services exist in the wild.
+Some only support specific operating systems or certain languages and for some you need to pay a monthly rate to use them.
+What all have in common: They have their own way to specify their setup file, which is most often written in YAML and stored at the top-level directory of your repo.
```
```suggestion
Each system has its own way to specify the setup. 
Often it is a YAML file stored at the top-level directory of your repo.
```

krlmlr commented on vignettes/advantages.Rmd:
```diff
@@ -0,0 +1,176 @@
+- Rather than dealing with all the CI differences, *tic* enables the specification of the complete workflow in an external R script file `tic.R`.
+- The commands specified in there will be understood on every CI service that is supported by *tic*.
+- Furthermore, you can [emulate all the stages and steps locally](advantages.html#less-dependency-on-external-ci-control-commands) by calling `tic()`.
+- You are independent from all the changes that are being made on the R setup for Travis or Appveyor ([more info]((advantages.html#less-dependency-on-external-ci-control-commands)).
```
```suggestion
- You are independent from all the changes that are being made to the R setup for Travis or Appveyor ([more info](advantages.html#less-dependency-on-external-ci-control-commands)).
```

krlmlr commented on vignettes/advantages.Rmd:
```diff
@@ -0,0 +1,176 @@
+We could list way more differences - but that's exactly the point when *tic* comes in!
+
+- Rather than dealing with all the CI differences, *tic* enables the specification of the complete workflow in an external R script file `tic.R`.
+- The commands specified in there will be understood on every CI service that is supported by *tic*.
```
```suggestion
- The commands specified in there will work the same way on every CI service that is supported by *tic*.
```

krlmlr commented on vignettes/advantages.Rmd:
```diff
@@ -0,0 +1,176 @@
+
+As a R package developer you are most likely used to call `devtools::check()` or directly `R CMD check` on your built package.
+Usually also Travis will run `R CMD check` to check the R package.
+*tic* instead uses  [*rcmdcheck*](https://github.com/r-lib/rcmdcheck), which can be seen as a wrapper around `R CMD build` and `R CMD check`.
```
```suggestion
*tic* instead uses  [*rcmdcheck*](https://github.com/r-lib/rcmdcheck), a wrapper around `R CMD build` and `R CMD check`.
```

krlmlr commented on vignettes/advantages.Rmd:
```diff
@@ -0,0 +1,176 @@
+
+- Coloring of important steps, simplifying the readability of the log
+- Enhanced and extended tracebacks of errors, especially errors in tests
+- The whole check process is returned in a R object, making it easier to inspect eventual errors/warnings
```
```suggestion
- The whole check process is returned in a R object, making it easier to inspect errors/warnings
```

krlmlr commented on vignettes/advantages.Rmd:
```diff
@@ -0,0 +1,176 @@
+
+## Caching of packages
+
+When using *tic* all dependencies (the ones of the package and of also the ones of all stages of the CI build) are installed in the "before_install" stage. 
```
```suggestion
When using *tic*, all dependencies (the ones of the package and also the ones of all stages of the CI build) are installed in the "before_install" stage. 
```

krlmlr commented on vignettes/advantages.Rmd:
```diff
@@ -0,0 +1,176 @@
+## Easier troubleshooting
+
+*tic* comes with the ability to emulate the CI run locally and debug potential problems. 
+This is possible because `tic.R` is a plain R file that will be sourced when calling `tic()`.
```
What is `tic()` ?

I think this can be kept very short, just mentioning that we can load the config with `load_from_file()`, and run it with `tic()`, and link to the corresponding vignette.

pat-s replied:
shortened heavily in https://github.com/ropenscilabs/tic/pull/80/commits/ad0716bc125a3133878b73ea64f9417ce1d90fdc.

pat-s commented on vignettes/advanced.Rmd:
```diff
@@ -11,203 +11,94 @@ vignette: >
-- No Pull Request
-- No CRON job ([CRON jobs](https://docs.travis-ci.com/user/cron-jobs/) are builds that are started automatically after no build was triggered after a specific time)
-- Only for builds that use the R "devel" version
+The conditioned execution of commands is based on environment variables.
```
I rephrased and shortened the paragraph and added generic example code blocks: https://github.com/ropenscilabs/tic/pull/80/commits/d65e58ddbef927d6b19e8a076a21cd4b8ea44dba.

pat-s commented on vignettes/advanced.Rmd:
````diff
@@ -11,203 +11,94 @@ vignette: >
-    add_step(step_build_pkgdown()) %>%
-    add_step(step_push_deploy())
+```{r eval = FALSE}
+if ((getRversion() == "release")) {
````
Ofc -> https://github.com/ropenscilabs/tic/pull/80/commits/73d2f0593bfc7819af8510379d56e4bb3b772ab2

pat-s commented on vignettes/advanced.Rmd:
```diff
@@ -11,203 +11,94 @@ vignette: >
 
-## Committing single files
+A special use case here would be to evaluate the R version inside an argument of `step_rcmdcheck()`.
+Here, we set `warnings_are_errors = TRUE` for `getRversion() >= "3.2")`.
```
Not existent anymore in new version.

pat-s commented on vignettes/advanced.Rmd:
````diff
@@ -221,20 +112,174 @@ See `?ci` for more information on which information can be extracted from this f
+load_from_file()
+```
+
+```{r eval=FALSE}
````
Did it for highlighting purposes in the Rmd file in RStudio.

krlmlr commented on vignettes/deployment.Rmd:
```diff
@@ -0,0 +1,188 @@
+To use this option with *tic*, specify `branch = "gh-pages", path = "docs"` in `step_push_deploy()`.
+As the site serving in this setting will use the root of the branch, the `gh-pages` branch will only consist of the public _pkgdown_ files used to serve the site. 
+
+### Conditional deployment
```
I still don't understand this section. I think we can point to Travis CI documentation for restricting deployment and for multi-stage build, and describe only "our" way here.

pat-s replied:
Rephrased it and focused more on the _tic_ way: https://github.com/ropenscilabs/tic/pull/80/commits/2a72bb9ffc51170db9b79de8b6595dacf3a766ec.

pat-s commented on vignettes/build-lifecycle.Rmd:
```diff
@@ -15,36 +15,27 @@ knitr::opts_chunk$set(
 The `after_deploy` stage will only be run if there the `deploy` stage was run before.
 
-`tic` also uses the stages approach and simply fowards all the commands that should be run in stage to the `tic.R` file where these will specified in an CI-agnostic way using R syntax.
+*tic* also uses the stages approach and forwards all action specifications that should be run in a stage to the `tic.R` file.
```
Should be a "transfers/redirects". 

Changed in https://github.com/ropenscilabs/tic/pull/80/commits/6bb209d47402c8ed530925d3f6bc4e406e79f62d.

pat-s commented on vignettes/custom-steps.Rmd:
````diff
@@ -65,17 +78,57 @@ res <- rcmdcheck::rcmdcheck(args = private$args)
+  HelloWorld$new()
+}
 ```
+
````
I wrote it in a more detailed way. Not sure if I got everything on point here. Also the permalinks are probably problematic for upcoming code changes.
https://github.com/ropenscilabs/tic/pull/80/commits/77139dba5f83bd32f6ebd35afc3f5a7b97d155ad

pat-s commented on vignettes/deployment.Rmd:
```diff
@@ -0,0 +1,188 @@
+
+## pkgdown deployment
+
+Today, it becomes more and more standard to have a _pkgdown_ site for a R package which presents all the well-written vignettes and function documentations in a nicely styled web version.
```
I rewrote the paragraph.

krlmlr commented on vignettes/custom-steps.Rmd:
```diff
@@ -33,7 +33,12 @@ We recommend following the same pattern for your custom steps.
+
+The `TicStep` class is the main class that is inherited by all `step_xxx` functions.
+It has three public methods.
+Two of them (`prepare()` and `run()`) must be overriden by the new step class.
```
`run()` needs to be overridden so that the step has an effect, the other two have suitable implementations in the base class which kick in when the step doesn't override.

krlmlr commented on vignettes/custom-steps.Rmd:
```diff
@@ -57,6 +62,22 @@ For example, the `run()` function of class `RCMDcheck` looks as follows:
+- All ssh steps defined in [`step-ssh.R`](https://github.com/ropenscilabs/tic/blob/master/R/steps-ssh.R)
+- [`step_do_push_deploy()`](https://github.com/ropenscilabs/tic/blob/457a5d259c6861e322220ac51a0436436e5f214b/R/steps-git.R#L187-L217)
+- [`step_push_deploy()`](https://github.com/ropenscilabs/tic/blob/457a5d259c6861e322220ac51a0436436e5f214b/R/steps-git.R#L319-L359)
+- [`step_write_text_file()`](https://github.com/ropenscilabs/tic/blob/457a5d259c6861e322220ac51a0436436e5f214b/R/steps-write-text-file.R#L1-L24)
```
One example is sufficient, the code chunk needs an explanation.

krlmlr commented on vignettes/custom-steps.Rmd:
````diff
@@ -94,14 +94,19 @@ If you are unfamiliar with `R6` classes, here is a short guidance how the argume
 
 ```{r eval = FALSE}
-step_rcmdcheck <- function(warnings_are_errors = TRUE, notes_are_errors = FALSE,
+step_rcmdcheck <- function(...,
````
Seems easier to just print the code of the function than to maintain an exact copy here.

krlmlr commented on vignettes/custom-steps.Rmd:
```diff
@@ -18,7 +18,8 @@ knitr::opts_chunk$set(
 The following vignette shows how new steps can be created and how *tic* defines them.
-Internally, all steps are defined as [R6](https://github.com/wch/R6) classes.
+Internally, all steps are defined using the [R6](https://github.com/wch/R6) class system.
+For more information see on object-oriented programming using R6 see [this chapter](https://adv-r.hadley.nz/r6.html) in [Advanced R](https://adv-r.hadley.nz/).
```
We're mentioning "method", "member" etc. in this section. I'm fine with that, but then an understanding of R6 should be a "formal" prerequisite for this chapter, and not a suggestion. Agree that this isn't the right place for an intro to R6.

krlmlr commented on vignettes/custom-steps.Rmd:
```diff
@@ -18,7 +18,8 @@ knitr::opts_chunk$set(
-Internally, all steps are defined as [R6](https://github.com/wch/R6) classes.
+Internally, all steps are defined using the [R6](https://github.com/wch/R6) class system.
+For more information see on object-oriented programming using R6 see [this chapter](https://adv-r.hadley.nz/r6.html) in [Advanced R](https://adv-r.hadley.nz/).
 A step is an environment-like (or list-like) object with named elements `check`, `prepare`, and `run`.
```
If we're focusing on R6, we can say that a step is a subclass of `TicStep`, and take it from there.

krlmlr commented on vignettes/deployment.Rmd:
```diff
@@ -165,7 +156,7 @@ if (Sys.getenv("id_rsa") != "") {
 }
 
-if (Sys.getenv("BUILD_PKGDOWN") != "") {
+if (isTRUE("BUILD_PKGDOWN")) {
```
``` r
isTRUE("true")
#> [1] FALSE
```

<sup>Created on 2018-11-27 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

krlmlr commented:
Thanks!
